beacon-chain/db/kv/state.go

```diff
@@ -3,24 +3,70 @@ package kv
 import (
 	"context"
 
-	"github.com/prysmaticlabs/prysm/beacon-chain/db/filters"
+	"github.com/boltdb/bolt"
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
 	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
 )
 
-// State retrieval by filter criteria.
-// TODO(#3164): Implement.
-func (k *Store) State(ctx context.Context, f *filters.QueryFilter) (*pb.BeaconState, error) {
-	return nil, nil
+// State returns the saved state using input block root.
+func (k *Store) State(ctx context.Context, blockRoot [32]byte) (*pb.BeaconState, error) {
+	var s *pb.BeaconState
+	err := k.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(stateBucket)
+		enc := bucket.Get(blockRoot[:])
+		if enc == nil {
+			return nil
+		}
+
+		var err error
+		s, err = createState(enc)
+
+		return err
+	})
+	return s, err
 }
 
-// HeadState returns the latest canonical state in eth2.
-// TODO(#3164): Implement.
+// HeadState returns the latest canonical state in beacon chain.
 func (k *Store) HeadState(ctx context.Context) (*pb.BeaconState, error) {
-	return nil, nil
+	var s *pb.BeaconState
+	err := k.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(blocksBucket)
+		headBlkRoot := bucket.Get(headBlockRootKey)
```

Comment:
Block root? Shouldn't these be state?

Reply:
+1

Reply:
This function is to retrieve the head state which will be retrieved using head block root

Reply:
see L38